File: sonarqube-companion-frontend/src/app/project/project-summary.ts
import {HealthStatus} from '../health/health-status';
import {Violations} from '../violations/violations';

export class ProjectSummary {

  name: string;
  key: string;
  serverId: string;
  healthStatus: HealthStatus;
  violations: Violations;
  serverUrl: string;

  constructor(data) {
    this.name = data.name;
    this.key = data.key;
    this.serverId = data.serverId;
    this.healthStatus = HealthStatus[data.healthStatus] as HealthStatus;
    this.violations = new Violations(data.violations || {});
    this.serverUrl = data.serverUrl;
  }

  get healthy(): boolean {
    return HealthStatus.HEALTHY === this.healthStatus;
  }

  get healthStatusString() {
    return this.healthStatus ? HealthStatus[this.healthStatus].toLowerCase() : '';
  }

}

File: sonarqube-companion-rest/src/main/java/pl/consdata/ico/sqcompanion/project/ProjectSummary.java
package pl.consdata.ico.sqcompanion.project;

import lombok.Builder;
import lombok.Data;
import pl.consdata.ico.sqcompanion.health.HealthStatus;
import pl.consdata.ico.sqcompanion.violation.Violations;

import java.util.List;
import java.util.Objects;
import java.util.function.ToIntFunction;

/**
 * @author gregorry
 */
@Data
@Builder
public class ProjectSummary {

    private String name;
    private String key;
    private String serverId;
    private String serverUrl;
    private HealthStatus healthStatus;
    private Violations violations;

    public static Violations summarizedViolations(final List<ProjectSummary> projectSummaries) {
        return Violations
                .builder()
                .blockers(sumProjectViolations(projectSummaries, Violations::getBlockers))
                .criticals(sumProjectViolations(projectSummaries, Violations::getCriticals))
                .majors(sumProjectViolations(projectSummaries, Violations::getMajors))
                .minors(sumProjectViolations(projectSummaries, Violations::getMinors))
                .infos(sumProjectViolations(projectSummaries, Violations::getInfos))
                .build();
    }

    public static int sumProjectViolations(final List<ProjectSummary> projectSummaries, final ToIntFunction<Violations> violationsExtractor) {
        return projectSummaries.stream().map(ProjectSummary::getViolations).filter(Objects::nonNull).mapToInt(violationsExtractor).sum();
    }

}
